File: backend/app/domain/flex_presets.py
"""加载 flex 预设树，并按当前块列表做 best-effort 适配。"""

from __future__ import annotations

import json
from functools import lru_cache
from typing import Any

from pydantic import BaseModel

from app.core.paths import SHARED_DIR
from app.domain.flex_layout import FlexContainer, FlexLeaf, FlexNode
from app.domain.flex_normalize import normalize

PRESETS_DIR = SHARED_DIR / "flex-presets"

# 预设叶子 block_id 暗示的类型偏好（用于匹配真实块）
_ROLE_TYPE_HINTS: dict[str, tuple[str, ...]] = {
    "title": ("text",),
    "kicker": ("text",),
    "lead": ("text",),
    "body": ("bullets", "text", "table", "cards"),
    "body_left": ("bullets", "text", "cards"),
    "body_right": ("bullets", "text", "cards"),
    "cards": ("cards",),
    "callout": ("callout",),
    "note": ("callout",),
    "source": ("callout",),
    "image": ("image", "chart"),
    "visual": ("image", "chart"),
    "kpi_1": ("kpi",),
    "kpi_2": ("kpi",),
    "kpi_3": ("kpi",),
    "kpi": ("kpi",),
}

_TYPE_TEXT_STYLE: dict[str, str | None] = {
    "text": "body",
    "bullets": "bullet",
    "kpi": None,
    "image": None,
    "chart": None,
    "table": None,
    "cards": None,
    "callout": None,
}


class FlexPreset(BaseModel):
    id: str
    name: str
    description: str = ""
    tree: FlexContainer
    expected_rects: dict[str, Any] | None = None


class BlockRef(BaseModel):
    """适配时只需要块 id 与类型。"""

    id: str
    type: str


@lru_cache(maxsize=4)
def load_presets() -> tuple[FlexPreset, ...]:
    if not PRESETS_DIR.is_dir():
        return ()
    presets: list[FlexPreset] = []
    for path in sorted(PRESETS_DIR.glob("*.json")):
        # 跳过 golden* 回归样本，不进入生产预设池
        if path.name.startswith("golden"):
            continue
        raw = json.loads(path.read_text(encoding="utf-8"))
        presets.append(FlexPreset.model_validate(raw))
    return tuple(presets)


def pick_preset_for_blocks(
    blocks: list[BlockRef],
    *,
    page_role: str | None = None,
    key_points: list[str] | None = None,
) -> FlexPreset | None:
    """按块类型组成与语义信号挑选最合适的预设。"""
    presets = list(load_presets())
    if not presets:
        return None
    types = {block.type for block in blocks}
    scored: list[tuple[int, FlexPreset]] = []
    for preset in presets:
        score = _score_preset(
            preset,
            types,
            len(blocks),
            page_role=page_role,
            key_points=key_points,
        )
        scored.append((score, preset))
    scored.sort(key=lambda item: (-item[0], item[1].id))
    return scored[0][1]


def adapt_preset_to_blocks(
    tree: FlexContainer,
    blocks: list[BlockRef],
) -> FlexContainer:
    """把预设树的叶子改写为给定块 id；多余叶子删除，未放入的块追加到根下。"""
    if not blocks:
        raise ValueError("至少需要一个内容块")

    remaining = list(blocks)
    adapted = _remap_node(tree, remaining)
    # 未匹配到的块挂到根容器末尾
    children = list(adapted.children)
    for block in remaining:
        children.append(
            FlexLeaf(
                id=f"leaf-{block.id}",
                block_id=block.id,
                grow=1.0,
                text_style=_text_style_for(block),
            )
        )
    result = adapted.model_copy(update={"children": children})
    return normalize(result)


def seed_layout_for_blocks(
    blocks: list[BlockRef],
    *,
    page_role: str | None = None,
    key_points: list[str] | None = None,
) -> FlexContainer:
    """从预设池选出一棵树并适配到当前块。"""
    preset = pick_preset_for_blocks(
        blocks, page_role=page_role, key_points=key_points
    )
    if preset is None:
        return normalize(
            FlexContainer(
                type="column",
                id="root",
                children=[
                    FlexLeaf(
                        id=f"leaf-{block.id}",
                        block_id=block.id,
                        grow=0.5 if block.type == "text" else 1.0,
                        text_style=_text_style_for(block),
                    )
                    for block in blocks
                ],
            )
        )
    return adapt_preset_to_blocks(preset.tree, blocks)


def alternate_preset_trees(
    blocks: list[BlockRef],
    *,
    exclude_tree: FlexContainer | None = None,
    limit: int = 1,
) -> list[FlexContainer]:
    """返回若干与当前树不同的预设适配结果。"""
    current_sig = _tree_signature(exclude_tree) if exclude_tree is not None else None
    results: list[FlexContainer] = []
    for preset in load_presets():
        adapted = adapt_preset_to_blocks(preset.tree, list(blocks))
        if current_sig is not None and _tree_signature(adapted) == current_sig:
            continue
        results.append(adapted)
        if len(results) >= limit:
            break
    return results


_STEP_KEYWORDS = ("步骤", "流程", "清单", "方法", "怎么做", "如何")
_TIMELINE_KEYWORDS = ("阶段", "时间", "里程碑", "路线", "演进", "历程", "节奏")


def _score_preset(
    preset: FlexPreset,
    types: set[str],
    block_count: int,
    *,
    page_role: str | None = None,
    key_points: list[str] | None = None,
) -> int:
    leaf_roles = _collect_roles(preset.tree)
    leaf_count = len(leaf_roles)
    preset_skins = _collect_presets(preset.tree)
    score = 0
    if "image" in types and any(role in ("image", "visual") for role in leaf_roles):
        score += 3
    if "chart" in types and any(role in ("image", "visual") for role in leaf_roles):
        score += 2
    if "kpi" in types and any(role.startswith("kpi") for role in leaf_roles):
        score += 3
    if "bullets" in types and any(role.startswith("body") for role in leaf_roles):
        score += 2
    if "cards" in types and any(role == "cards" or "cards" in role for role in leaf_roles):
        score += 4
    if "callout" in types and any(
        role in ("callout", "note", "source") for role in leaf_roles
    ):
        score += 3
    # 叶子数接近加分；多块页面偏好更密预设
    score -= abs(leaf_count - block_count) * 2
    if block_count >= 4 and leaf_count >= 4:
        score += 2
    if block_count >= 5 and leaf_count >= 5:
        score += 2

    # 语义信号：步骤/时间类文案加权对应皮肤预设
    blob = " ".join(key_points or [])
    if any(word in blob for word in _STEP_KEYWORDS) and "numbered_steps" in preset_skins:
        score += 4
    if any(word in blob for word in _TIMELINE_KEYWORDS) and "timeline" in preset_skins:
        score += 4
    if page_role == "summary" and "callout" in types:
        score += 1
    if page_role == "content" and "kicker" in leaf_roles and "text" in types:
        score += 1
    return score


def _collect_presets(node: FlexNode) -> set[str]:
    if isinstance(node, FlexLeaf):
        return set()
    found: set[str] = set()
    if node.preset is not None:
        found.add(node.preset)
    for child in node.children:
        found |= _collect_presets(child)
    return found


def _collect_roles(node: FlexNode) -> list[str]:
    if isinstance(node, FlexLeaf):
        return [node.block_id]
    roles: list[str] = []
    for child in node.children:
        roles.extend(_collect_roles(child))
    return roles


def _remap_node(node: FlexNode, remaining: list[BlockRef]) -> FlexNode:
    if isinstance(node, FlexLeaf):
        match = _claim_block(node.block_id, remaining)
        if match is None:
            # 空 id 叶子由父级过滤
            return node.model_copy(update={"block_id": ""})
        return FlexLeaf(
            id=f"leaf-{match.id}",
            block_id=match.id,
            grow=node.grow,
            text_style=node.text_style or _text_style_for(match),
        )

    children: list[FlexNode] = []
    for child in node.children:
        remapped = _remap_node(child, remaining)
        if isinstance(remapped, FlexLeaf) and remapped.block_id == "":
            continue
        if isinstance(remapped, FlexContainer) and not remapped.children:
            continue
        children.append(remapped)
    return node.model_copy(update={"children": children})


def _claim_block(role: str, remaining: list[BlockRef]) -> BlockRef | None:
    hints = _ROLE_TYPE_HINTS.get(role, ())
    for preferred in hints:
        for index, block in enumerate(remaining):
            if block.type == preferred:
                return remaining.pop(index)
    # 角色名与类型对不上时，按常见语义再试：title→text，其余取第一个
    if role == "title":
        for index, block in enumerate(remaining):
            if block.type == "text":
                return remaining.pop(index)
    if remaining:
        return remaining.pop(0)
    return None


def _text_style_for(block: BlockRef) -> str | None:
    lower = block.id.lower()
    if block.type == "text":
        if "kicker" in lower or "eyebrow" in lower:
            return "caption"
        if "lead" in lower or "subtitle" in lower:
            return "subtitle"
        if "title" in lower or lower.endswith("-title") or lower == "title":
            return "title"
    return _TYPE_TEXT_STYLE.get(block.type)


def _tree_signature(tree: FlexContainer) -> str:
    """结构签名：忽略叶子 id，保留方向与叶子顺序类型占位。"""

    def walk(node: FlexNode) -> Any:
        if isinstance(node, FlexLeaf):
            return ("block",)
        return (node.type, [walk(child) for child in node.children], node.ratios)

    return json.dumps(walk(tree), ensure_ascii=False)

File: java_backend/src/main/java/com/aippt/domain/flex/FlexPresets.java
package com.aippt.domain.flex;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.HashSet;
import java.util.Iterator;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.stream.Stream;

import org.springframework.stereotype.Component;

import com.aippt.shared.config.RepoPaths;
import com.aippt.shared.json.JsonMapperHolder;
import com.fasterxml.jackson.databind.JsonNode;

@Component
public class FlexPresets {

    private static final Map<String, List<String>> ROLE_TYPE_HINTS = Map.ofEntries(
            Map.entry("title", List.of("text")),
            Map.entry("kicker", List.of("text")),
            Map.entry("lead", List.of("text")),
            Map.entry("body", List.of("bullets", "text", "table", "cards")),
            Map.entry("body_left", List.of("bullets", "text", "cards")),
            Map.entry("body_right", List.of("bullets", "text", "cards")),
            Map.entry("cards", List.of("cards")),
            Map.entry("callout", List.of("callout")),
            Map.entry("note", List.of("callout")),
            Map.entry("source", List.of("callout")),
            Map.entry("image", List.of("image", "chart")),
            Map.entry("visual", List.of("image", "chart")),
            Map.entry("kpi_1", List.of("kpi")),
            Map.entry("kpi_2", List.of("kpi")),
            Map.entry("kpi_3", List.of("kpi")),
            Map.entry("kpi", List.of("kpi"))
    );
    private static final Map<String, String> TYPE_TEXT_STYLE = Map.of(
            "text", "body",
            "bullets", "bullet"
    );
    private static final List<String> STEP_KEYWORDS = List.of("步骤", "流程", "清单", "方法", "怎么做", "如何");
    private static final List<String> TIMELINE_KEYWORDS = List.of("阶段", "时间", "里程碑", "路线", "演进", "历程", "节奏");

    public record BlockRef(String id, String type) {
    }

    public record FlexPreset(String id, String name, String description, FlexContainer tree) {
    }

    private volatile List<FlexPreset> presets;

    public FlexContainer seedLayoutForBlocks(List<BlockRef> blocks, String pageRole, List<String> keyPoints) {
        FlexPreset preset = pickPreset(blocks, pageRole, keyPoints);
        if (preset == null) {
            List<FlexNode> leaves = new ArrayList<>();
            for (BlockRef block : blocks) {
                leaves.add(FlexLeaf.of(
                        "leaf-" + block.id(),
                        block.id(),
                        "text".equals(block.type()) ? 0.5 : 1.0,
                        textStyleFor(block)
                ));
            }
            return FlexNormalize.normalize(FlexContainer.column("root", leaves, 16.0, 1.0));
        }
        return adaptPreset(preset.tree(), blocks);
    }

    public List<FlexContainer> alternatePresetTrees(List<BlockRef> blocks, FlexContainer excludeTree, int limit) {
        String currentSig = excludeTree == null ? null : treeSignature(excludeTree);
        List<FlexContainer> results = new ArrayList<>();
        for (FlexPreset preset : presets()) {
            FlexContainer adapted = adaptPreset(preset.tree(), new ArrayList<>(blocks));
            if (currentSig != null && currentSig.equals(treeSignature(adapted))) {
                continue;
            }
            results.add(adapted);
            if (results.size() >= limit) {
                break;
            }
        }
        return results;
    }

    private static String treeSignature(FlexContainer tree) {
        try {
            return com.aippt.shared.json.JsonMapperHolder.MAPPER.writeValueAsString(FlexTrees.signature(tree));
        } catch (Exception ex) {
            return String.valueOf(FlexTrees.signature(tree));
        }
    }

    private FlexPreset pickPreset(List<BlockRef> blocks, String pageRole, List<String> keyPoints) {
        List<FlexPreset> loaded = presets();
        if (loaded.isEmpty()) {
            return null;
        }
        Set<String> types = new HashSet<>();
        for (BlockRef block : blocks) {
            types.add(block.type());
        }
        FlexPreset best = null;
        int bestScore = Integer.MIN_VALUE;
        for (FlexPreset preset : loaded) {
            int score = score(preset, types, blocks.size(), pageRole, keyPoints);
            if (best == null || score > bestScore || (score == bestScore && preset.id().compareTo(best.id()) < 0)) {
                best = preset;
                bestScore = score;
            }
        }
        return best;
    }

    FlexContainer adaptPreset(FlexContainer tree, List<BlockRef> blocks) {
        if (blocks.isEmpty()) {
            throw new IllegalArgumentException("至少需要一个内容块");
        }
        List<BlockRef> remaining = new ArrayList<>(blocks);
        FlexNode adapted = remap(tree, remaining);
        List<FlexNode> children = new ArrayList<>(((FlexContainer) adapted).children());
        for (BlockRef block : remaining) {
            children.add(FlexLeaf.of("leaf-" + block.id(), block.id(), 1.0, textStyleFor(block)));
        }
        return FlexNormalize.normalize(((FlexContainer) adapted).withChildren(children));
    }

    private FlexNode remap(FlexNode node, List<BlockRef> remaining) {
        if (node instanceof FlexLeaf leaf) {
            BlockRef match = claim(leaf.blockId(), remaining);
            if (match == null) {
                return leaf.withBlockId("", leaf.id());
            }
            return FlexLeaf.of(
                    "leaf-" + match.id(),
                    match.id(),
                    leaf.grow(),
                    leaf.textStyle() == null ? textStyleFor(match) : leaf.textStyle()
            );
        }
        FlexContainer container = (FlexContainer) node;
        List<FlexNode> children = new ArrayList<>();
        for (FlexNode child : container.children()) {
            FlexNode remapped = remap(child, remaining);
            if (remapped instanceof FlexLeaf leaf && leaf.blockId().isEmpty()) {
                continue;
            }
            if (remapped instanceof FlexContainer nested && nested.children().isEmpty()) {
                continue;
            }
            children.add(remapped);
        }
        return container.withChildren(children);
    }

    private BlockRef claim(String role, List<BlockRef> remaining) {
        List<String> hints = ROLE_TYPE_HINTS.getOrDefault(role, List.of());
        for (String preferred : hints) {
            Iterator<BlockRef> iterator = remaining.iterator();
            while (iterator.hasNext()) {
                BlockRef block = iterator.next();
                if (preferred.equals(block.type())) {
                    iterator.remove();
                    return block;
                }
            }
        }
        if ("title".equals(role)) {
            Iterator<BlockRef> iterator = remaining.iterator();
            while (iterator.hasNext()) {
                BlockRef block = iterator.next();
                if ("text".equals(block.type())) {
                    iterator.remove();
                    return block;
                }
            }
        }
        if (!remaining.isEmpty()) {
            return remaining.remove(0);
        }
        return null;
    }

    private int score(FlexPreset preset, Set<String> types, int blockCount, String pageRole, List<String> keyPoints) {
        List<String> leafRoles = FlexTrees.collectRoles(preset.tree());
        Set<String> skins = FlexTrees.collectPresets(preset.tree());
        int leafCount = leafRoles.size();
        int score = 0;
        if (types.contains("image") && leafRoles.stream().anyMatch(role -> "image".equals(role) || "visual".equals(role))) {
            score += 3;
        }
        if (types.contains("chart") && leafRoles.stream().anyMatch(role -> "image".equals(role) || "visual".equals(role))) {
            score += 2;
        }
        if (types.contains("kpi") && leafRoles.stream().anyMatch(role -> role.startsWith("kpi"))) {
            score += 3;
        }
        if (types.contains("bullets") && leafRoles.stream().anyMatch(role -> role.startsWith("body"))) {
            score += 2;
        }
        if (types.contains("cards") && leafRoles.stream().anyMatch(role -> "cards".equals(role) || role.contains("cards"))) {
            score += 4;
        }
        if (types.contains("callout") && leafRoles.stream().anyMatch(role -> Set.of("callout", "note", "source").contains(role))) {
            score += 3;
        }
        score -= Math.abs(leafCount - blockCount) * 2;
        if (blockCount >= 4 && leafCount >= 4) {
            score += 2;
        }
        if (blockCount >= 5 && leafCount >= 5) {
            score += 2;
        }
        String blob = keyPoints == null ? "" : String.join(" ", keyPoints);
        if (containsAny(blob, STEP_KEYWORDS) && skins.contains("numbered_steps")) {
            score += 4;
        }
        if (containsAny(blob, TIMELINE_KEYWORDS) && skins.contains("timeline")) {
            score += 4;
        }
        if ("summary".equals(pageRole) && types.contains("callout")) {
            score += 1;
        }
        if ("content".equals(pageRole) && leafRoles.contains("kicker") && types.contains("text")) {
            score += 1;
        }
        return score;
    }

    private static String textStyleFor(BlockRef block) {
        String lower = block.id().toLowerCase();
        if ("text".equals(block.type())) {
            if (lower.contains("kicker") || lower.contains("eyebrow")) {
                return "caption";
            }
            if (lower.contains("lead") || lower.contains("subtitle")) {
                return "subtitle";
            }
            if (lower.contains("title") || lower.endsWith("-title") || "title".equals(lower)) {
                return "title";
            }
        }
        return TYPE_TEXT_STYLE.get(block.type());
    }

    private static boolean containsAny(String blob, List<String> words) {
        for (String word : words) {
            if (blob.contains(word)) {
                return true;
            }
        }
        return false;
    }

    private List<FlexPreset> presets() {
        if (presets == null) {
            synchronized (this) {
                if (presets == null) {
                    presets = load();
                }
            }
        }
        return presets;
    }

    private static List<FlexPreset> load() {
        Path dir = RepoPaths.flexPresetsDir();
        if (!Files.isDirectory(dir)) {
            return List.of();
        }
        List<FlexPreset> loaded = new ArrayList<>();
        try (Stream<Path> files = Files.list(dir)) {
            files.filter(path -> path.getFileName().toString().endsWith(".json"))
                    .filter(path -> !path.getFileName().toString().startsWith("golden"))
                    .sorted()
                    .forEach(path -> loaded.add(read(path)));
        } catch (IOException ex) {
            throw new IllegalStateException("无法读取 flex 预设目录 " + dir, ex);
        }
        return List.copyOf(loaded);
    }

    private static FlexPreset read(Path path) {
        try {
            JsonNode raw = JsonMapperHolder.MAPPER.readTree(path.toFile());
            return new FlexPreset(
                    raw.path("id").asText(),
                    raw.path("name").asText(),
                    raw.path("description").asText(""),
                    FlexTrees.parse(raw.get("tree"))
            );
        } catch (IOException ex) {
            throw new IllegalStateException("无法读取 flex 预设 " + path, ex);
        }
    }
}
